File: _test/fixtures/provision.py
#!/usr/bin/env python3
"""Apply users.json to one LDAP backend.

Runs INSIDE the matching LDAP container via the compose post_start
lifecycle hook (see _test/docker-compose.yml). The same script handles
both backends; pick which one with --backend.

Three classes:
    Provisioner          - abstract base, driver loop and shared helpers
    OpenLDAPProvisioner  - implements per-entity ops via ldapadd / ldapmodify
    SambaProvisioner     - implements them via samba-tool / ldbmodify

users.json is pure data: no per-backend tags, no connection info, no
conditions. Each subclass reads its own connection settings from
environment variables (set on the matching service in docker-compose.yml).

Both run() invocations are idempotent: each backend probes for the
canonical 'alice' entry and short-circuits if she's already there. On
success a sentinel file (/tmp/provisioned) is touched, which the
service's compose-level healthcheck reads — so `docker compose up
--wait` blocks until provisioning is actually done.
"""

import argparse
import base64
import hashlib
import json
import os
import shutil
import subprocess
import sys
import time
from abc import ABC, abstractmethod
from pathlib import Path

SPEC_PATH = Path(__file__).parent / "users.json"
SENTINEL = Path("/tmp/provisioned")


class Provisioner(ABC):
    def __init__(self, spec):
        self.spec = spec

    def run(self):
        name = type(self).__name__
        self.wait_until_ready()
        if self.is_already_provisioned():
            print(f"[{name}] already provisioned, skipping", flush=True)
            SENTINEL.touch()
            return
        print(f"[{name}] provisioning...", flush=True)

        self.ensure_root()
        self.apply_password_policy()
        for g in self.spec["groups"]:
            self.add_group(g)
        for g in self.spec["groups"]:
            for inner in g.get("contains", []):
                self.add_group_to_group(g["name"], inner)
        for u in self.spec["users"]:
            self.add_user(u)
        for u in self.spec["users"]:
            for grp in u.get("memberOf", []):
                self.add_user_to_group(u["uid"], grp)
        if "padding" in self.spec:
            self.add_padding_users(self.spec["padding"])
        self.finalize()
        SENTINEL.touch()
        print(f"[{name}] done", flush=True)

    @abstractmethod
    def wait_until_ready(self): ...
    @abstractmethod
    def is_already_provisioned(self): ...
    @abstractmethod
    def ensure_root(self): ...
    @abstractmethod
    def apply_password_policy(self): ...
    @abstractmethod
    def add_group(self, g): ...
    @abstractmethod
    def add_user(self, u): ...
    @abstractmethod
    def add_user_to_group(self, uid, group): ...
    @abstractmethod
    def add_group_to_group(self, outer, inner): ...
    @abstractmethod
    def add_padding_users(self, padding): ...
    @abstractmethod
    def finalize(self): ...

    @staticmethod
    def env(key):
        v = os.environ.get(key)
        if not v:
            raise RuntimeError(f"Missing required env var: {key}")
        return v

    @staticmethod
    def run_cmd(cmd, stdin=None, allow_fail=False):
        r = subprocess.run(cmd, input=stdin, text=True, capture_output=True)
        if r.returncode != 0 and not allow_fail:
            sys.stderr.write(f"\nCommand failed (exit {r.returncode}): {cmd}\n")
            if stdin is not None:
                sys.stderr.write(f"stdin:\n{stdin}\n")
            if r.stdout:
                sys.stderr.write(f"stdout:\n{r.stdout}\n")
            if r.stderr:
                sys.stderr.write(f"stderr:\n{r.stderr}\n")
            sys.exit(1)
        return r


class OpenLDAPProvisioner(Provisioner):
    def __init__(self, spec):
        super().__init__(spec)
        self.host = self.env("PROVISION_HOST")
        self.port = int(self.env("PROVISION_PORT"))
        self.bind_dn = self.env("PROVISION_BIND_DN")
        self.bind_pw = self.env("PROVISION_BIND_PW")
        self.base = self.env("PROVISION_BASE")
        self.people_ou = self.env("PROVISION_PEOPLE_OU")
        self.groups_ou = self.env("PROVISION_GROUPS_OU")
        self.next_uid = 1000
        self.next_gid = 5000
        self._ensure_ldap_utils()

    def _ensure_ldap_utils(self):
        # osixia/openldap doesn't always ship ldap-utils. Install on
        # demand so we have ldapsearch / ldapadd / ldapmodify available.
        if shutil.which("ldapadd"):
            return
        print("Installing ldap-utils...", flush=True)
        env = {**os.environ, "DEBIAN_FRONTEND": "noninteractive"}
        subprocess.run(["apt-get", "update", "-qq"], check=True, env=env)
        subprocess.run(["apt-get", "install", "-y", "-qq", "ldap-utils"],
                       check=True, env=env)

    def wait_until_ready(self):
        print("[OpenLDAPProvisioner] waiting for slapd...", flush=True)
        deadline = time.time() + 120
        while time.time() < deadline:
            if self._entry_exists(self.base):
                return
            time.sleep(0.5)
        raise RuntimeError("slapd did not become responsive within 120s")

    def is_already_provisioned(self):
        return self._entry_exists(f"uid=alice,{self.people_ou}")

    def ensure_root(self):
        for dn in (self.people_ou, self.groups_ou):
            if self._entry_exists(dn):
                continue
            rdn_attr, rdn_val = dn.split(',', 1)[0].split('=', 1)
            self._ldap_add(
                f"dn: {dn}\n"
                f"objectClass: organizationalUnit\n"
                f"{rdn_attr}: {rdn_val}\n"
            )

    def apply_password_policy(self):
        pass  # not modelled for slapd in this fixture

    def add_group(self, g):
        self.next_gid += 1
        print(f"  + group {g['name']}", flush=True)
        self._ldap_add(
            f"dn: cn={g['name']},{self.groups_ou}\n"
            f"objectClass: posixGroup\n"
            f"cn: {g['name']}\n"
            f"gidNumber: {self.next_gid}\n"
        )

    def add_user(self, u):
        self.next_uid += 1
        cn = f"{u['givenName']} {u['sn']}"
        ldif = (
            f"dn: uid={u['uid']},{self.people_ou}\n"
            f"objectClass: inetOrgPerson\n"
            f"objectClass: posixAccount\n"
            f"uid: {u['uid']}\n"
            f"cn: {cn}\n"
            f"sn: {u['sn']}\n"
            f"givenName: {u['givenName']}\n"
        )
        if "mail" in u:
            ldif += f"mail: {u['mail']}\n"
        if "mobile" in u:
            ldif += f"mobile: {u['mobile']}\n"
        ldif += (
            f"uidNumber: {self.next_uid}\n"
            f"gidNumber: {self.next_uid}\n"
            f"homeDirectory: /home/{u['uid']}\n"
            f"userPassword: {_ssha(u['password'])}\n"
        )
        print(f"  + user {u['uid']}", flush=True)
        self._ldap_add(ldif)

    def add_user_to_group(self, uid, group):
        self._ldap_modify(
            f"dn: cn={group},{self.groups_ou}\n"
            f"changetype: modify\n"
            f"add: memberUid\n"
            f"memberUid: {uid}\n"
        )

    def add_group_to_group(self, outer, inner):
        # posixGroup uses memberUid (RFC 2307) and has no nesting concept.
        # The spec carries 'contains' for AD's sake; on slapd it's a no-op.
        pass

    def add_padding_users(self, p):
        print(f"  + padding users (1..{p['count']})", flush=True)
        for i in range(1, p["count"] + 1):
            uid = p["uid_format"] % i
            given = p["givenName_format"] % i
            self.next_uid += 1
            self._ldap_add(
                f"dn: uid={uid},{self.people_ou}\n"
                f"objectClass: inetOrgPerson\n"
                f"objectClass: posixAccount\n"
                f"uid: {uid}\n"
                f"cn: {given} {p['sn']}\n"
                f"sn: {p['sn']}\n"
                f"givenName: {given}\n"
                f"uidNumber: {self.next_uid}\n"
                f"gidNumber: {self.next_uid}\n"
                f"homeDirectory: /home/{uid}\n"
                f"userPassword: {_ssha(p['password'])}\n"
            )
            for m in p.get("memberships", []):
                if i <= m["first_n"]:
                    self.add_user_to_group(uid, m["group"])
            if i % 50 == 0:
                print(f"    ...{i}/{p['count']}", flush=True)

    def finalize(self):
        pass

    def _entry_exists(self, dn):
        r = self.run_cmd([
            "ldapsearch", "-x",
            "-H", f"ldap://{self.host}:{self.port}",
            "-D", self.bind_dn, "-w", self.bind_pw,
            "-b", dn, "-s", "base", "-LLL", "(objectClass=*)", "dn",
        ], allow_fail=True)
        return r.returncode == 0

    def _ldap_add(self, ldif):
        self.run_cmd([
            "ldapadd", "-x",
            "-H", f"ldap://{self.host}:{self.port}",
            "-D", self.bind_dn, "-w", self.bind_pw,
        ], stdin=ldif)

    def _ldap_modify(self, ldif):
        self.run_cmd([
            "ldapmodify", "-x",
            "-H", f"ldap://{self.host}:{self.port}",
            "-D", self.bind_dn, "-w", self.bind_pw,
        ], stdin=ldif)


class SambaProvisioner(Provisioner):
    SAM_LDB = "/var/lib/samba/private/sam.ldb"
    TLS_DIR = "/var/lib/samba/private/tls"

    def __init__(self, spec):
        super().__init__(spec)
        self.users_dn = self.env("PROVISION_USERS_DN")

    def wait_until_ready(self):
        # samba-tool only responds once the AD-DC has finished its
        # own first-boot domain provisioning. That takes 30-90s.
        print("[SambaProvisioner] waiting for samba-tool (slow first boot)...",
              flush=True)
        deadline = time.time() + 300
        while time.time() < deadline:
            r = subprocess.run(["samba-tool", "user", "list"],
                               stdout=subprocess.DEVNULL,
                               stderr=subprocess.DEVNULL)
            if r.returncode == 0:
                return
            time.sleep(2)
        raise RuntimeError("samba did not become responsive within 300s")

    def is_already_provisioned(self):
        r = subprocess.run(["samba-tool", "user", "show", "alice"],
                           stdout=subprocess.DEVNULL,
                           stderr=subprocess.DEVNULL)
        return r.returncode == 0

    def ensure_root(self):
        pass  # CN=Users,DC=… is created by samba's first-boot provisioning

    def apply_password_policy(self):
        days = self.spec.get("password_policy", {}).get("max_pwd_age_days")
        if days is None:
            return
        self.run_cmd([
            "samba-tool", "domain", "passwordsettings", "set",
            f"--max-pwd-age={days}",
        ])

    def add_group(self, g):
        print(f"  + group {g['name']}", flush=True)
        self.run_cmd(["samba-tool", "group", "add", g["name"]])

    def add_user(self, u):
        print(f"  + user {u['uid']}", flush=True)
        cmd = [
            "samba-tool", "user", "create", u["uid"], u["password"],
            f"--given-name={u['givenName']}",
            f"--surname={u['sn']}",
            "--use-username-as-cn",
        ]
        if "mail" in u:
            cmd.append(f"--mail-address={u['mail']}")
        self.run_cmd(cmd)

        attrs = {"displayName": f"{u['givenName']} {u['sn']}"}
        if "mobile" in u:
            attrs["mobile"] = u["mobile"]
        if u.get("passwordNeverExpires"):
            # NORMAL_ACCOUNT 512 | DONT_EXPIRE_PASSWD 65536 = 66048
            attrs["userAccountControl"] = "66048"
        self._ldb_modify(f"CN={u['uid']},{self.users_dn}", attrs)

    def add_user_to_group(self, uid, group):
        self.run_cmd(["samba-tool", "group", "addmembers", group, uid])

    def add_group_to_group(self, outer, inner):
        self.run_cmd(["samba-tool", "group", "addmembers", outer, inner])

    def add_padding_users(self, p):
        print(f"  + padding users (1..{p['count']})", flush=True)
        for i in range(1, p["count"] + 1):
            uid = p["uid_format"] % i
            given = p["givenName_format"] % i
            self.run_cmd([
                "samba-tool", "user", "create", uid, p["password"],
                f"--given-name={given}",
                f"--surname={p['sn']}",
                "--use-username-as-cn",
            ])
            for m in p.get("memberships", []):
                if i <= m["first_n"]:
                    self.run_cmd([
                        "samba-tool", "group", "addmembers",
                        m["group"], uid,
                    ])
            if i % 50 == 0:
                print(f"    ...{i}/{p['count']}", flush=True)

    def finalize(self):
        # Samba's auto-generated TLS cert is issued for the DC's FQDN
        # (dc1.example.local). The plugin connects to localhost:7389
        # with STARTTLS — FreeDSx's hostname check rejects the handshake
        # even with validate=self. Reissue with a subjectAltName covering
        # localhost and 127.0.0.1, then HUP samba so it reloads.
        print("  + regenerate TLS cert with localhost SAN", flush=True)
        os.makedirs(self.TLS_DIR, exist_ok=True)
        self.run_cmd([
            "openssl", "req", "-x509", "-newkey", "rsa:2048", "-nodes",
            "-keyout", f"{self.TLS_DIR}/key.pem",
            "-out", f"{self.TLS_DIR}/cert.pem",
            "-days", "365",
            "-subj", "/CN=localhost",
            "-addext", "subjectAltName=DNS:localhost,DNS:dc1.example.local,IP:127.0.0.1",
        ])
        shutil.copy(f"{self.TLS_DIR}/cert.pem", f"{self.TLS_DIR}/ca.pem")
        os.chmod(f"{self.TLS_DIR}/key.pem", 0o600)
        subprocess.run(["pkill", "-HUP", "samba"], check=False)

    def _ldb_modify(self, dn, attrs):
        parts = [f"dn: {dn}\nchangetype: modify\n"]
        sep = ""
        for name, value in attrs.items():
            parts.append(f"{sep}replace: {name}\n{name}: {value}\n")
            sep = "-\n"
        self.run_cmd(["ldbmodify", "-H", self.SAM_LDB], stdin="".join(parts))


def _ssha(pw):
    # Deterministic salt so the LDIF is reproducible across runs.
    salt = hashlib.sha1(f"ssha-salt:{pw}".encode()).digest()[:4]
    h = hashlib.sha1(pw.encode() + salt).digest()
    return "{SSHA}" + base64.b64encode(h + salt).decode()


def _strip_comments(v):
    if isinstance(v, dict):
        return {k: _strip_comments(vv) for k, vv in v.items()
                if not (isinstance(k, str) and k.startswith("_comment"))}
    if isinstance(v, list):
        return [_strip_comments(x) for x in v]
    return v


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--backend", required=True, choices=["openldap", "samba"])
    args = p.parse_args()

    with open(SPEC_PATH) as f:
        spec = _strip_comments(json.load(f))

    cls = {
        "openldap": OpenLDAPProvisioner,
        "samba": SambaProvisioner,
    }[args.backend]
    cls(spec).run()


if __name__ == "__main__":
    main()

File: _test/fixtures/provision.php
#!/usr/bin/env php
<?php

/**
 * Apply users.json to both LDAP backends.
 *
 * Three classes:
 *   Provisioner          — abstract base. Driver loop and shared helpers.
 *   OpenLDAPProvisioner  — implements the per-entity methods via ldapadd /
 *                          ldapmodify against the openldap service.
 *   SambaProvisioner     — implements them via docker-exec → samba-tool /
 *                          ldbmodify against the samba container.
 *
 * The spec is plain data — no backend tags, no connection info, no
 * per-backend conditions. Each subclass reads its own connection
 * settings from environment variables (set on the provisioner service
 * in docker-compose.yml) and decides how to express the spec's fields
 * against its backend (e.g. posixAccount uidNumber is auto-assigned;
 * passwordNeverExpires becomes a userAccountControl override on AD
 * and a no-op on slapd).
 *
 * Both run() invocations are idempotent: each backend probes for the
 * canonical 'alice' entry and short-circuits if she's already there,
 * so re-running this script against a populated fixture is a no-op.
 */

const SPEC_PATH = __DIR__ . '/users.json';

abstract class Provisioner
{
    protected array $spec;

    public function __construct(array $spec)
    {
        $this->spec = $spec;
    }

    public function run(): void
    {
        $name = (new ReflectionClass($this))->getShortName();
        $this->waitUntilReady();
        if ($this->isAlreadyProvisioned()) {
            echo "[{$name}] already provisioned, skipping\n";
            return;
        }
        echo "[{$name}] provisioning...\n";

        $this->ensureRoot();
        $this->applyPasswordPolicy();

        foreach ($this->spec['groups'] as $g) {
            $this->addGroup($g);
        }
        foreach ($this->spec['groups'] as $g) {
            foreach ($g['contains'] ?? [] as $inner) {
                $this->addGroupToGroup($g['name'], $inner);
            }
        }
        foreach ($this->spec['users'] as $u) {
            $this->addUser($u);
        }
        foreach ($this->spec['users'] as $u) {
            foreach ($u['memberOf'] ?? [] as $group) {
                $this->addUserToGroup($u['uid'], $group);
            }
        }
        if (isset($this->spec['padding'])) {
            $this->addPaddingUsers($this->spec['padding']);
        }
        $this->finalize();
        echo "[{$name}] done\n";
    }

    abstract protected function waitUntilReady(): void;
    abstract protected function isAlreadyProvisioned(): bool;
    abstract protected function ensureRoot(): void;
    abstract protected function applyPasswordPolicy(): void;
    abstract protected function addGroup(array $g): void;
    abstract protected function addUser(array $u): void;
    abstract protected function addUserToGroup(string $uid, string $group): void;
    abstract protected function addGroupToGroup(string $outer, string $inner): void;
    abstract protected function addPaddingUsers(array $padding): void;
    abstract protected function finalize(): void;

    protected function env(string $key): string
    {
        $v = getenv($key);
        if ($v === false || $v === '') {
            throw new RuntimeException("Missing required env var: {$key}");
        }
        return $v;
    }

    protected function runCmd(array $cmd, ?string $stdin = null): string
    {
        $proc = proc_open($cmd, [
            0 => $stdin !== null ? ['pipe', 'r'] : ['file', '/dev/null', 'r'],
            1 => ['pipe', 'w'],
            2 => ['pipe', 'w'],
        ], $pipes);
        if (!is_resource($proc)) {
            throw new RuntimeException('proc_open failed: ' . implode(' ', $cmd));
        }
        if ($stdin !== null) {
            fwrite($pipes[0], $stdin);
            fclose($pipes[0]);
        }
        $out = stream_get_contents($pipes[1]);
        $err = stream_get_contents($pipes[2]);
        fclose($pipes[1]);
        fclose($pipes[2]);
        $code = proc_close($proc);
        if ($code !== 0) {
            fwrite(STDERR, "\nCommand failed (exit {$code}): " . implode(' ', $cmd) . "\n");
            if ($stdin !== null) fwrite(STDERR, "stdin:\n{$stdin}\n");
            if ($out !== '')     fwrite(STDERR, "stdout:\n{$out}\n");
            if ($err !== '')     fwrite(STDERR, "stderr:\n{$err}\n");
            exit(1);
        }
        return $out;
    }

    protected function runCmdStatus(array $cmd): int
    {
        $proc = proc_open($cmd, [
            0 => ['file', '/dev/null', 'r'],
            1 => ['file', '/dev/null', 'w'],
            2 => ['file', '/dev/null', 'w'],
        ], $pipes);
        return is_resource($proc) ? proc_close($proc) : 1;
    }
}


class OpenLDAPProvisioner extends Provisioner
{
    private string $host;
    private int    $port;
    private string $bindDn;
    private string $bindPw;
    private string $peopleOu;
    private string $groupsOu;
    private int    $nextUidNumber = 1000;
    private int    $nextGidNumber = 5000;

    public function __construct(array $spec)
    {
        parent::__construct($spec);
        $this->host     = $this->env('OPENLDAP_HOST');
        $this->port     = (int) $this->env('OPENLDAP_PORT');
        $this->bindDn   = $this->env('OPENLDAP_BIND_DN');
        $this->bindPw   = $this->env('OPENLDAP_BIND_PW');
        $this->peopleOu = $this->env('OPENLDAP_PEOPLE_OU');
        $this->groupsOu = $this->env('OPENLDAP_GROUPS_OU');
    }

    protected function waitUntilReady(): void
    {
        echo "[OpenLDAPProvisioner] waiting for slapd...\n";
        $deadline = time() + 120;
        while (time() < $deadline) {
            if ($this->entryExists($this->env('OPENLDAP_BASE'))) {
                return;
            }
            usleep(500_000);
        }
        throw new RuntimeException("slapd did not become responsive within 120s");
    }

    protected function isAlreadyProvisioned(): bool
    {
        return $this->entryExists("uid=alice,{$this->peopleOu}");
    }

    protected function ensureRoot(): void
    {
        // osixia/openldap creates the dc=… root from LDAP_DOMAIN but no
        // sub-OUs; seed the two we plant entries under.
        foreach ([$this->peopleOu, $this->groupsOu] as $dn) {
            if ($this->entryExists($dn)) continue;
            [$attr, $val] = explode('=', explode(',', $dn, 2)[0], 2);
            $this->ldapAdd("dn: {$dn}\nobjectClass: organizationalUnit\n{$attr}: {$val}\n");
        }
    }

    protected function applyPasswordPolicy(): void
    {
        // Not modelled for slapd in this fixture.
    }

    protected function addGroup(array $g): void
    {
        $gid = ++$this->nextGidNumber;
        echo "  + group {$g['name']}\n";
        $this->ldapAdd(
            "dn: cn={$g['name']},{$this->groupsOu}\n"
            . "objectClass: posixGroup\n"
            . "cn: {$g['name']}\n"
            . "gidNumber: {$gid}\n"
        );
    }

    protected function addUser(array $u): void
    {
        $uidNumber = ++$this->nextUidNumber;
        $cn = "{$u['givenName']} {$u['sn']}";
        $ldif  = "dn: uid={$u['uid']},{$this->peopleOu}\n";
        $ldif .= "objectClass: inetOrgPerson\n";
        $ldif .= "objectClass: posixAccount\n";
        $ldif .= "uid: {$u['uid']}\n";
        $ldif .= "cn: {$cn}\n";
        $ldif .= "sn: {$u['sn']}\n";
        $ldif .= "givenName: {$u['givenName']}\n";
        if (isset($u['mail']))   $ldif .= "mail: {$u['mail']}\n";
        if (isset($u['mobile'])) $ldif .= "mobile: {$u['mobile']}\n";
        $ldif .= "uidNumber: {$uidNumber}\n";
        $ldif .= "gidNumber: {$uidNumber}\n";
        $ldif .= "homeDirectory: /home/{$u['uid']}\n";
        $ldif .= 'userPassword: ' . $this->sshaHash($u['password']) . "\n";
        echo "  + user {$u['uid']}\n";
        $this->ldapAdd($ldif);
    }

    protected function addUserToGroup(string $uid, string $group): void
    {
        $this->ldapModify(
            "dn: cn={$group},{$this->groupsOu}\n"
            . "changetype: modify\n"
            . "add: memberUid\n"
            . "memberUid: {$uid}\n"
        );
    }

    protected function addGroupToGroup(string $outer, string $inner): void
    {
        // posixGroup uses memberUid (RFC 2307) and doesn't model nested
        // groups. The spec carries 'contains' for AD's sake; on slapd
        // it's silently ignored.
    }

    protected function addPaddingUsers(array $p): void
    {
        echo "  + padding users (1..{$p['count']})\n";
        for ($i = 1; $i <= $p['count']; $i++) {
            $uid       = sprintf($p['uid_format'], $i);
            $given     = sprintf($p['givenName_format'], $i);
            $uidNumber = ++$this->nextUidNumber;
            $ldif  = "dn: uid={$uid},{$this->peopleOu}\n";
            $ldif .= "objectClass: inetOrgPerson\n";
            $ldif .= "objectClass: posixAccount\n";
            $ldif .= "uid: {$uid}\n";
            $ldif .= "cn: {$given} {$p['sn']}\n";
            $ldif .= "sn: {$p['sn']}\n";
            $ldif .= "givenName: {$given}\n";
            $ldif .= "uidNumber: {$uidNumber}\n";
            $ldif .= "gidNumber: {$uidNumber}\n";
            $ldif .= "homeDirectory: /home/{$uid}\n";
            $ldif .= 'userPassword: ' . $this->sshaHash($p['password']) . "\n";
            $this->ldapAdd($ldif);
            foreach ($p['memberships'] ?? [] as $m) {
                if ($i <= $m['first_n']) {
                    $this->addUserToGroup($uid, $m['group']);
                }
            }
            if ($i % 50 === 0) echo "    ...{$i}/{$p['count']}\n";
        }
    }

    protected function finalize(): void
    {
        // nothing
    }

    private function entryExists(string $dn): bool
    {
        return 0 === $this->runCmdStatus([
            'ldapsearch', '-x',
            '-H', "ldap://{$this->host}:{$this->port}",
            '-D', $this->bindDn, '-w', $this->bindPw,
            '-b', $dn, '-s', 'base', '-LLL', '(objectClass=*)', 'dn',
        ]);
    }

    private function ldapAdd(string $ldif): void
    {
        $this->runCmd([
            'ldapadd', '-x',
            '-H', "ldap://{$this->host}:{$this->port}",
            '-D', $this->bindDn, '-w', $this->bindPw,
        ], $ldif);
    }

    private function ldapModify(string $ldif): void
    {
        $this->runCmd([
            'ldapmodify', '-x',
            '-H', "ldap://{$this->host}:{$this->port}",
            '-D', $this->bindDn, '-w', $this->bindPw,
        ], $ldif);
    }

    private function sshaHash(string $pw): string
    {
        // Deterministic salt so the LDIF is reproducible.
        $salt = substr(hash('sha1', "ssha-salt:{$pw}", true), 0, 4);
        return '{SSHA}' . base64_encode(sha1($pw . $salt, true) . $salt);
    }
}


class SambaProvisioner extends Provisioner
{
    private string $container;
    private string $usersDn;
    private const LDB_PATH = '/var/lib/samba/private/sam.ldb';

    public function __construct(array $spec)
    {
        parent::__construct($spec);
        $this->container = $this->env('SAMBA_CONTAINER');
        $this->usersDn   = $this->env('SAMBA_USERS_DN');
    }

    protected function waitUntilReady(): void
    {
        echo "[SambaProvisioner] waiting for samba-tool (this is slow, samba self-provisions the domain first)...\n";
        $deadline = time() + 300;
        while (time() < $deadline) {
            $ok = 0 === $this->runCmdStatus([
                'docker', 'exec', $this->container,
                'samba-tool', 'user', 'list',
            ]);
            if ($ok) return;
            sleep(2);
        }
        throw new RuntimeException("samba did not become responsive within 300s");
    }

    protected function isAlreadyProvisioned(): bool
    {
        return 0 === $this->runCmdStatus([
            'docker', 'exec', $this->container,
            'samba-tool', 'user', 'show', 'alice',
        ]);
    }

    protected function ensureRoot(): void
    {
        // CN=Users,DC=… is created by samba's own provisioning at first
        // boot; nothing for us to seed.
    }

    protected function applyPasswordPolicy(): void
    {
        if (!isset($this->spec['password_policy']['max_pwd_age_days'])) {
            return;
        }
        $days = $this->spec['password_policy']['max_pwd_age_days'];
        $this->dockerExec([
            'samba-tool', 'domain', 'passwordsettings', 'set',
            "--max-pwd-age={$days}",
        ]);
    }

    protected function addGroup(array $g): void
    {
        echo "  + group {$g['name']}\n";
        $this->dockerExec(['samba-tool', 'group', 'add', $g['name']]);
    }

    protected function addUser(array $u): void
    {
        echo "  + user {$u['uid']}\n";
        $cmd = [
            'samba-tool', 'user', 'create', $u['uid'], $u['password'],
            '--given-name=' . $u['givenName'],
            '--surname='    . $u['sn'],
            '--use-username-as-cn',
        ];
        if (isset($u['mail'])) $cmd[] = '--mail-address=' . $u['mail'];
        $this->dockerExec($cmd);

        // Drop a single ldbmodify with all the post-create attribute
        // overrides; samba-tool exposes only a subset of attrs as flags.
        $attrs = ['displayName' => "{$u['givenName']} {$u['sn']}"];
        if (isset($u['mobile'])) {
            $attrs['mobile'] = $u['mobile'];
        }
        if (!empty($u['passwordNeverExpires'])) {
            // 512 NORMAL_ACCOUNT | 65536 DONT_EXPIRE_PASSWD = 66048
            $attrs['userAccountControl'] = '66048';
        }
        $this->ldbModify("CN={$u['uid']},{$this->usersDn}", $attrs);
    }

    protected function addUserToGroup(string $uid, string $group): void
    {
        $this->dockerExec(['samba-tool', 'group', 'addmembers', $group, $uid]);
    }

    protected function addGroupToGroup(string $outer, string $inner): void
    {
        $this->dockerExec(['samba-tool', 'group', 'addmembers', $outer, $inner]);
    }

    protected function addPaddingUsers(array $p): void
    {
        echo "  + padding users (1..{$p['count']}, this takes a minute)\n";
        for ($i = 1; $i <= $p['count']; $i++) {
            $uid   = sprintf($p['uid_format'], $i);
            $given = sprintf($p['givenName_format'], $i);
            $this->dockerExec([
                'samba-tool', 'user', 'create', $uid, $p['password'],
                "--given-name={$given}",
                "--surname={$p['sn']}",
                '--use-username-as-cn',
            ]);
            foreach ($p['memberships'] ?? [] as $m) {
                if ($i <= $m['first_n']) {
                    $this->dockerExec(['samba-tool', 'group', 'addmembers', $m['group'], $uid]);
                }
            }
            if ($i % 50 === 0) echo "    ...{$i}/{$p['count']}\n";
        }
    }

    protected function finalize(): void
    {
        echo "  + regenerate TLS cert with localhost SAN\n";
        $this->dockerExec(['/regen-tls.sh']);
    }

    private function dockerExec(array $cmd, ?string $stdin = null): void
    {
        $argv = ['docker', 'exec'];
        if ($stdin !== null) $argv[] = '-i';
        $argv[] = $this->container;
        array_push($argv, ...$cmd);
        $this->runCmd($argv, $stdin);
    }

    private function ldbModify(string $dn, array $attrs): void
    {
        $ldif = "dn: {$dn}\nchangetype: modify\n";
        $first = true;
        foreach ($attrs as $name => $value) {
            if (!$first) $ldif .= "-\n";
            $ldif .= "replace: {$name}\n{$name}: {$value}\n";
            $first = false;
        }
        $this->dockerExec(['ldbmodify', '-H', self::LDB_PATH], $ldif);
    }
}


// ---------------------------------------------------------------------------

$spec = json_decode(file_get_contents(SPEC_PATH), true, 512, JSON_THROW_ON_ERROR);

// Strip _comment keys (allowed in users.json for human notes)
$strip = function ($v) use (&$strip) {
    if (!is_array($v)) return $v;
    $out = [];
    foreach ($v as $k => $vv) {
        if (is_string($k) && str_starts_with($k, '_comment')) continue;
        $out[$k] = $strip($vv);
    }
    return $out;
};
$spec = $strip($spec);

(new OpenLDAPProvisioner($spec))->run();
(new SambaProvisioner($spec))->run();

echo "All backends provisioned.\n";
